### PR TITLE
Lua cache improvement

### PR DESCRIPTION
**Pull request by bouda1**
The goal here is to add methods to the broker cache provided by the Lua module.

New methods are:
get_hostgroup_name(hg_id)
get_hostgroups(host_id)
get_servicegroup_name(sg_id)
get_servicegroups(host_id, service_id)

They are all tested in unit tests.

### CODE REVIEW

**Comment by bouda1** [comment]
This branch is fixed so that we do not talk about bam objects here.
A new branch based on this one is kept with bam objects, but this is another story.

File: lua/src/macro_cache.cc [hunk 1]
```diff
@@ -108,6 +109,36 @@ QString const& macro_cache::get_host_name(unsigned int host_id) const {
...
+ *  @param[in] id  The id of the host group.
+ *
+ *  @return             The name of the host group.
+ */
+QString const& macro_cache::get_host_group_name(unsigned int id) const {
+  QHash<unsigned int, neb::host_group>::const_iterator
+    found(_host_groups.find(id));
+
+  if (found == _host_groups.end())
+    throw (exceptions::msg()
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: How are exceptions handled by the Lua interpreter ?

File: lua/src/macro_cache.cc [hunk 2]
```diff
@@ -188,6 +257,34 @@ void macro_cache::_process_host(neb::host const& h) {
...
+ *  Process a host group event.
+ *
+ *  @param hg  The event.
+ */
+void macro_cache::_process_host_group(neb::host_group const& hg) {
+  logging::debug(logging::medium)
+    << "macro_cache: process host group --- "
+    << "host group id = " << hg.id
+    << " ; name = " << hg.name;
+  _host_groups[hg.id] = hg;
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: The *enabled* flag should be handled to check if group has to be added or removed from the cache.

File: lua/src/macro_cache.cc [hunk 3]
```diff
@@ -188,6 +257,34 @@ void macro_cache::_process_host(neb::host const& h) {
...
 }
 
+/**
+ *  Process a host group event.
+ *
+ *  @param hg  The event.
+ */
+void macro_cache::_process_host_group(neb::host_group const& hg) {
+  logging::debug(logging::medium)
+    << "macro_cache: process host group --- "
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Could you normalize the log message ? The descriptor should be _lua:_ and properties are usually provided within parenthesis (check the SQL module for more examples).

File: lua/src/macro_cache.cc [hunk 4]
```diff
@@ -188,6 +257,34 @@ void macro_cache::_process_host(neb::host const& h) {
...
+}
+
+/**
+ *  Process a host group member event.
+ *
+ *  @param hgm  The event.
+ */
+void macro_cache::_process_host_group_member(
+       neb::host_group_member const& hgm) {
+  logging::debug(logging::medium)
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Log message should be updated.

File: lua/src/macro_cache.cc [hunk 5]
```diff
@@ -188,6 +257,34 @@ void macro_cache::_process_host(neb::host const& h) {
...
+    << "macro_cache: process host group member --- "
+    << "host id = " << hgm.host_id
+    << "host group id = " << hgm.group_id
+    << "host group name = " << hgm.group_name;
+  _host_group_members.insert(hgm.host_id, hgm);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: *enabled* should be handled (see above).

File: lua/src/macro_cache.cc [hunk 6]
```diff
@@ -202,6 +299,35 @@ void macro_cache::_process_service(neb::service const& s) {
   _services[qMakePair(s.host_id, s.service_id)] = s;
 }
 
+/**
+ *  Process a service group event.
+ *
+ *  @param sg  The event.
+ */
+void macro_cache::_process_service_group(neb::service_group const& sg) {
+  logging::debug(logging::medium)
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Please update log message (see above).

File: lua/src/macro_cache.cc [hunk 7]
```diff
@@ -202,6 +299,35 @@ void macro_cache::_process_service(neb::service const& s) {
...
+    << "macro_cache: process service group --- "
+    << "service group id = " << sg.id
+    << " ; name = " << sg.name;
+  _service_groups[sg.id] = sg;
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: _enabled_ should be handled (see above).

File: lua/src/macro_cache.cc [hunk 8]
```diff
@@ -202,6 +299,35 @@ void macro_cache::_process_service(neb::service const& s) {
...
+ */
+void macro_cache::_process_service_group_member(
+       neb::service_group_member const& sgm) {
+  logging::debug(logging::medium)
+    << "macro_cache: process service group member --- "
+    << "host id = " << sgm.host_id
+    << "service id = " << sgm.service_id
+    << "service group id = " << sgm.group_id
+    << "service group name = " << sgm.group_name;
+  _service_group_members.insert(qMakePair(sgm.host_id, sgm.service_id), sgm);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: _enabled_ should be handled (see above).

File: lua/src/macro_cache.cc [hunk 9]
```diff
@@ -202,6 +299,35 @@ void macro_cache::_process_service(neb::service const& s) {
...
+}
+
+/**
+ *  Process a service group member event.
+ *
+ *  @param sgm  The event.
+ */
+void macro_cache::_process_service_group_member(
+       neb::service_group_member const& sgm) {
+  logging::debug(logging::medium)
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Please update log message (see above).

File: lua/src/macro_cache.cc (hunk 1 above)
Comment (bouda1) [reply lua/src/macro_cache.cc]: They are not.
That's why there is a try /  catch in l_broker_cache functions, so that we can return an error.

Comment (ganoze) [reply lua/src/macro_cache.cc]: Perfect :-)

File: lua/src/macro_cache.cc [hunk 10]
```diff
@@ -278,11 +277,13 @@ void macro_cache::_process_host_group(neb::host_group const& hg) {
...
-    << "macro_cache: process host group member --- "
-    << "host id = " << hgm.host_id
-    << "host group id = " << hgm.group_id
-    << "host group name = " << hgm.group_name;
-  _host_group_members.insert(hgm.host_id, hgm);
+    << "lua: processing host group member "
+    << " (group_name: '" << hgm.group_name << "', group_id: " << hgm.group_id
+    << ", host_id: " << hgm.host_id << ")";
+  if (hgm.enabled)
+    _host_group_members.insert(hgm.host_id, hgm);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: You need to check if membership is already existing to prevent double insertion (or change container type).

File: lua/src/macro_cache.cc [hunk 11]
```diff
@@ -182,37 +251,94 @@ void macro_cache::_process_instance(instance_broadcast const& in) {
...
+void macro_cache::_process_host_group_member(
+       neb::host_group_member const& hgm) {
+  logging::debug(logging::medium)
+    << "lua: processing host group member "
+    << " (group_name: '" << hgm.group_name << "', group_id: " << hgm.group_id
+    << ", host_id: " << hgm.host_id << ")";
+  if (hgm.enabled)
+    _host_group_members.insert(hgm.host_id, hgm);
+  else
+    _host_group_members.remove(hgm.host_id);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Removal should only concern this specific membership (host / hostgroup pair), not all memberships of the host.

File: lua/src/macro_cache.cc [hunk 12]
```diff
@@ -182,37 +251,94 @@ void macro_cache::_process_instance(instance_broadcast const& in) {
...
+ */
+void macro_cache::_process_service_group_member(
+       neb::service_group_member const& sgm) {
+  logging::debug(logging::medium)
+    << "lua: processing service group member "
+    << " (group_name: '" << sgm.group_name << "', group_id: " << sgm.group_id
+    << ", host_id: " << sgm.host_id
+    << ", service_id: " << sgm.service_id << ")";
+  if (sgm.enabled)
+    _service_group_members.insert(qMakePair(sgm.host_id, sgm.service_id), sgm);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: My comments about host group membership insertion / removal also apply to service group memberships.

File: lua/inc/com/centreon/broker/lua/macro_cache.hh [hunk 13]
```diff
@@ -58,9 +68,14 @@ namespace         lua {
 
     void           _process_instance(instance_broadcast const& in);
```
Comment (ganoze) [inline lua/inc/com/centreon/broker/lua/macro_cache.hh]: The __process_instance()_ method should clean cache. Will this be available in another PR ?

Comment (bouda1) [reply lua/inc/com/centreon/broker/lua/macro_cache.hh]: So this is also an issue in influxdb code...

File: lua/src/macro_cache.cc (hunk 10 above)
Comment (bouda1) [reply lua/src/macro_cache.cc]: My bad, I forgot the "multi"... :-(

File: lua/inc/com/centreon/broker/lua/macro_cache.hh (hunk 13 above)
Comment (ganoze) [reply lua/inc/com/centreon/broker/lua/macro_cache.hh]: Yes indeed the issue is in InfluxDB and Graphite modules too.

When receiving an _instance_broadcast_ event, all items *related to this instance* should be discarded. This means that hash tables should not be cleared entirely but instead objects related to the target instance should be removed. With the current data structure this will likely be inefficient though.

File: lua/src/macro_cache.cc [hunk 14]
```diff
@@ -172,7 +243,67 @@ void macro_cache::write(misc::shared_ptr<io::data> const& data) {
...
+         it(_host_groups.begin()),
+         end(_host_groups.end());
+       it != end; ) {
+    if (it->poller_id == poller_id)
+      it = _host_groups.erase(it);
+    else
+      ++it;
+  }
+
+  for (QHash<unsigned int,
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: In fact it is not possible to remove host groups based on their poller ID. The same host group can span on multiple instances. Unfortunately, I have no solution here.

File: lua/src/macro_cache.cc [hunk 15]
```diff
@@ -182,37 +313,96 @@ void macro_cache::_process_instance(instance_broadcast const& in) {
...
+ *
+ *  @param hg  The event.
+ */
+void macro_cache::_process_host_group(neb::host_group const& hg) {
+  logging::debug(logging::medium)
+    << "lua: processing host group '" << hg.name << "' of id " << hg.id;
+  if (hg.enabled)
+    _host_groups[hg.id] = hg;
+  else
+    _host_groups.remove(hg.id);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Can't do that on groups. The same group can span on multiple instances.

File: lua/src/macro_cache.cc [hunk 16]
```diff
@@ -182,37 +313,96 @@ void macro_cache::_process_instance(instance_broadcast const& in) {
...
+ *
+ *  @param sg  The event.
+ */
+void macro_cache::_process_service_group(neb::service_group const& sg) {
+  logging::debug(logging::medium)
+    << "lua: processing service group '" << sg.name << "' of id " << sg.id;
+  if (sg.enabled)
+    _service_groups[sg.id] = sg;
+  else
+    _service_groups.remove(sg.id);
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Can't do that on groups. The same group can span on multiple instances.

File: lua/src/macro_cache.cc [hunk 17]
```diff
@@ -182,37 +313,96 @@ void macro_cache::_process_instance(instance_broadcast const& in) {
...
 /**
  *  Process an index mapping event.
  *
  *  @param im  The event.
  */
 void macro_cache::_process_index_mapping(storage::index_mapping const& im) {
   logging::debug(logging::medium)
-    << "macro_cache: process index mapping --- "
-    << "index id = " << im.index_id;
+    << "lua: processing index mapping (index_id: " << im.index_id << ")";
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Is it possible to have _host_id_ and _service_id_ in message ?

File: lua/src/macro_cache.cc [hunk 18]
```diff
@@ -222,8 +412,7 @@ void macro_cache::_process_index_mapping(storage::index_mapping const& im) {
  */
 void macro_cache::_process_metric_mapping(storage::metric_mapping const& mm) {
   logging::debug(logging::medium)
-    << "macro_cache: process metric mapping --- "
-    << "index id = " << mm.index_id;
+    << "lua: processing metric mapping (index_id: " << mm.index_id << ")";
```
Comment (ganoze) [inline lua/src/macro_cache.cc]: Is it possible to have _metric_id_ in message ?
